### PR TITLE
[DevEx] VS Code remote-containers setup

### PR DESCRIPTION
**Pull request by electronjoe**
# Summary

This PR populates a functional .devcontainer root folder within the magma repo - enabling ~full function c/c++ **build, unit test and tooling e.g. Clang-Tidy, Clang-Format** in the following very-low-friction settings. It is a goal of mine to work towards E2E containerized test support, such that we can run E2E tests here as well.

- GH Code Spaces
  - Automatically builds container and mounts repo within
- Local vscode development
  - Install Microsoft's VS Code [Remote - Containers](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers) Extension
  - Simply hit F1 and select `Remote-Containers: Clone Repository in Containers Volume...`
  - Paste in a GH URL to a repo branch, or a PR
  - It will build the docker container, mount your repo as a docker volume, voila

The following configuration is automatically applied and available.

- Container environment
  - Software
    - All AGW c/c++ dependencies
    - Ubuntu 20.04 based container
    - GCC 9.3
    - Clang 11 tool suite (incl clang-tidy, clang-format)
  - Magma Source
    - Is mounted in container via a Docker Volume automatically
    - Resides at working directory of terminal (/workspaces/magma/)
- VS Code Configuration
  - Extensions auto-installed
    - ms-vscode.cpptools (c/c++ syntax highlighting / compilation)
    - CoenraadS.bracket-pair-colorizer-2 (color-matched parenthesis)
    - xaver.clang-format (clang-format support - see warning below)
    - notskm.clang-tidy (clang-tidy findings directly in vs code with ~squiggles)
    - eamodio.gitlens (fantastic git overlays of knowledge)
    - yzhang.markdown-all-in-one (markdown support)
    - christian-kohler.path-intellisense (path auto-complete in code)
    - vscodevim.vim (vim key bindings)
    - vscode-icons-team.vscode-icons (improved iconography)
    - ms-vsliveshare.vsliveshare (live-share support with colleagues)
    - ms-azuretools.vscode-docker (docker support, needed for local dev)
  - Settings auto-applied
    - c/c++ include paths necessary for compilation / analysis
    - clang-tidy include paths necessary for analysis
    - (these path sets are a work in progress, contributions welcome)

# Using

## With GH Code Spaces

To use in GH Code Spaces - simply open up a GH Codespace for our repo (from a branch containing this commit, select the Green code icon, drop down and select `Open with Codespaces`). Note you can work entirely in the browser if you like, or you can attach your local VS Code to the Codespaces environment.

## Locally

See `Setup` in `.devcontainer/README.md`.

# Sharp Edges

Beware that our clang-format configuration, applied with clang-format 11, results in many modifications to our codebase formatting. This will need to be resolved soon as we are moving on towards Ubuntu 20.04 and can expect Clang-11 in whatever build environment we are running. My 2 cents is that we would be better served using a standard clang-format config e.g. Googles - which I expect is better specified and does not result in motion over clang versions.

# TODO

## Add Docker Registry

Presently the configuration uses no docker registry - so the Dockerfile is built from scratch (or cache if local) every environment. This takes ~40 minutes today. Easy enough to fix with help from e.g. @tmdzk and his team.

## Add any packages / support needed for Golang / Cloud dev

Presently this Dockerfile focuses on AGW c/c++ development. I do believe vs-code can provide environments per task - or we could create one docker container with all our needs (once a registry is in use, this will have ~minimal impact?).

## Reduce Dockerfile Duplication

I'm not sure how feasible this will be - but **much** of `.devcontainer/Dockerfile` is shared with `lte/gateway/docker/mme/Dockerfile.ubuntu20.04`.  But there are substantial base image differences as well - so I'm not sure how well Docker re-use mechanisms will work here. It would be much better to share common library install / setup stages between the two - so we do not duplicate content.

# Test Plan

From this branch in my personal magma repo, checked out in GH Code Spaces and confirmed Dockerfile built.  Opened terminal and successfully did the following.

```
cd lte/gateway
make test_oai
```

Signed-off-by: Scott Harrison Moeller <smoeller@fb.com>

### CODE REVIEW

File: .devcontainer/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+CMD [ "sleep", "infinity" ]
+
+ARG GIT_PROXY
+ARG FEATURES=mme_oai
+ENV MAGMA_ROOT=/workspaces/magma
+ENV BUILD_TYPE=Debug
+ENV C_BUILD=/workspaces/magma/build/c
+ENV DEBIAN_FRONTEND=noninteractive
+
+RUN echo "Install general purpouse packages" && \
```
Comment (tmdzk) [inline .devcontainer/Dockerfile]: typo

File: .devcontainer/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+ARG ENABLE_NONROOT_DOCKER="true"
+# [Option] Use the OSS Moby CLI instead of the licensed Docker CLI
+ARG USE_MOBY="true"
+
+# Install needed packages and setup non-root user. Use a separate RUN statement to add your
+# own dependencies. A user of "automatic" attempts to reuse an user ID if one already exists.
+ARG USERNAME=automatic
+ARG USER_UID=1000
+ARG USER_GID=$USER_UID
+RUN echo $(ls -al .)
```
Comment (tmdzk) [inline .devcontainer/Dockerfile]: Can you remove it?

File: .devcontainer/library-scripts/common-debian.sh [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+INSTALL_ZSH=${1:-"true"}
+USERNAME=${2:-"automatic"}
+USER_UID=${3:-"automatic"}
+USER_GID=${4:-"automatic"}
+UPGRADE_PACKAGES=${5:-"true"}
+INSTALL_OH_MYS=${6:-"true"}
+
+set -e
+
+if [ "$(id -u)" -ne 0 ]; then
```
Comment (tmdzk) [inline .devcontainer/library-scripts/common-debian.sh:20]: Ohh that's elegant didn't know about that one.

**Review by tmdzk — APPROVED** [review]
Love it! Approved but a few minor comments.

**Comment by ShubhamTatvamasi** [comment]
@electronjoe we don't have docker permissions inside the container.

Screenshot from 2021-03-27 15-00-07

**Comment by electronjoe** [comment]
> @electronjoe we don't have docker permissions inside the container.
> 
> Screenshot from 2021-03-27 15-00-07

I haven't run into use cases (yet) for docker commands inside the container - can you give me some examples? I just haven't run into this yet.

**Comment by electronjoe** [comment]
@ShubhamTatvamasi I'm going to merge this already-sizeable PR and we can absolutely apply any fixups (e.g. new permissions to vscode, or removal of vscode user... etc) on top.

**Pull request merged**